File: frontend/src/lib/analytics.ts
import { apiRequest } from "@/lib/api/request";
import { readSessionFromStorage } from "@/lib/auth-session";
import {
  hasCurrentConsent,
  readConsentRecord,
  type ConsentDecisions,
} from "@/lib/consent";
import { getFrontendRuntimeConfig } from "@/lib/runtime-config";

declare global {
  interface Window {
    _paq?: Array<unknown[]>;
  }
}

export type AnalyticsSurface = "public" | "app" | "admin";

export type AnalyticsFeatureArea =
  | "documents"
  | "chat"
  | "evaluations"
  | "settings"
  | "connectors"
  | "public"
  | "dashboard"
  | "admin";

export type AnalyticsEventName =
  | "activation.signup_completed"
  | "activation.organization_created"
  | "activation.first_upload"
  | "activation.first_indexed_document"
  | "activation.first_question"
  | "activation.first_cited_answer"
  | "feature.documents.viewed"
  | "feature.documents.uploaded"
  | "feature.documents.indexed"
  | "feature.dashboard.viewed"
  | "feature.chat.viewed"
  | "feature.chat.question_submitted"
  | "feature.chat.answer_rendered"
  | "feature.chat.citation_opened"
  | "feature.chat.retrieval_diagnostics_viewed"
  | "feature.evaluations.viewed"
  | "feature.settings.viewed"
  | "feature.connectors.viewed"
  | "feature.public_page.viewed";

export type AnalyticsEventPayload = {
  surface: AnalyticsSurface;
  route?: string;
  pageKey?: string;
  featureArea?: AnalyticsFeatureArea;
  entityId?: string;
  entityType?: string;
  status?: string;
  method?: string;
  count?: number;
  citationCount?: number;
  hasCitations?: boolean;
  locale?: string;
  source?: string;
  dedupeKey?: string;
};

type StoredAnalyticsRecord = {
  key: string;
  recordedAt: number;
};

type MatomoConfig = {
  url: string;
  siteId: string;
};

type TrackOptions = {
  dedupeOnce?: boolean;
  sendToBackend?: boolean;
};

const ANALYTICS_STORAGE_KEY = "rudix.analytics.v1";
const ANALYTICS_BACKEND_EVENT_VERSION = 1;

function trimToNull(value: string | null | undefined): string | null {
  if (!value) {
    return null;
  }
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : null;
}

function getMatomoConfig(): MatomoConfig | null {
  const config = getFrontendRuntimeConfig();
  if (!config.features.analyticsEnabled) {
    return null;
  }

  if (!config.analytics.matomoUrl || !config.analytics.matomoSiteId) {
    return null;
  }

  return {
    url: config.analytics.matomoUrl,
    siteId: config.analytics.matomoSiteId,
  };
}

function readStoredRecords(): StoredAnalyticsRecord[] {
  if (typeof window === "undefined") {
    return [];
  }

  try {
    const raw = window.localStorage.getItem(ANALYTICS_STORAGE_KEY);
    if (!raw) {
      return [];
    }
    const parsed = JSON.parse(raw) as unknown;
    if (!Array.isArray(parsed)) {
      return [];
    }

    return parsed.filter((item): item is StoredAnalyticsRecord => {
      if (typeof item !== "object" || item === null) return false;
      const candidate = item as Record<string, unknown>;
      return (
        typeof candidate.key === "string" &&
        candidate.key.length > 0 &&
        typeof candidate.recordedAt === "number"
      );
    });
  } catch {
    return [];
  }
}

function persistStoredRecords(records: StoredAnalyticsRecord[]): void {
  if (typeof window === "undefined") {
    return;
  }

  try {
    window.localStorage.setItem(ANALYTICS_STORAGE_KEY, JSON.stringify(records));
  } catch {
    // Ignore storage failures.
  }
}

function rememberOnce(key: string): boolean {
  if (typeof window === "undefined") {
    return true;
  }

  const records = readStoredRecords();
  if (records.some((record) => record.key === key)) {
    return false;
  }

  const nextRecords = [...records, { key, recordedAt: Date.now() }].slice(-200);
  persistStoredRecords(nextRecords);
  return true;
}

function canTrack(): boolean {
  return (
    hasCurrentConsent() && getFrontendRuntimeConfig().features.analyticsEnabled
  );
}

function normalizePathname(value: string | undefined): string | null {
  const trimmed = trimToNull(value);
  if (!trimmed) {
    return null;
  }
  return trimmed.startsWith("/") ? trimmed : `/${trimmed}`;
}

function safeName(value: string | null | undefined): string {
  return trimToNull(value) ?? "";
}

function sanitizePayload(
  payload: AnalyticsEventPayload,
): Record<string, unknown> {
  return {
    surface: payload.surface,
    route: normalizePathname(payload.route),
    page_key: trimToNull(payload.pageKey),
    feature_area: payload.featureArea ?? null,
    entity_id: trimToNull(payload.entityId),
    entity_type: trimToNull(payload.entityType),
    status: trimToNull(payload.status),
    method: trimToNull(payload.method),
    count: payload.count ?? null,
    citation_count: payload.citationCount ?? null,
    has_citations: payload.hasCitations ?? null,
    locale: trimToNull(payload.locale),
    source: trimToNull(payload.source),
    dedupe_key: trimToNull(payload.dedupeKey),
  };
}

function eventKey(
  eventName: AnalyticsEventName,
  payload: AnalyticsEventPayload,
): string {
  return [
    eventName,
    payload.surface,
    normalizePathname(payload.route) ?? "",
    trimToNull(payload.pageKey) ?? "",
    trimToNull(payload.entityId) ?? "",
    trimToNull(payload.dedupeKey) ?? "",
  ].join("|");
}

function resolvePageViewEventName(
  surface: AnalyticsSurface,
  featureArea: AnalyticsFeatureArea,
): AnalyticsEventName {
  if (surface === "public") {
    return "feature.public_page.viewed";
  }

  switch (featureArea) {
    case "documents":
      return "feature.documents.viewed";
    case "dashboard":
      return "feature.dashboard.viewed";
    case "chat":
      return "feature.chat.viewed";
    case "evaluations":
      return "feature.evaluations.viewed";
    case "settings":
      return "feature.settings.viewed";
    case "connectors":
      return "feature.connectors.viewed";
    default:
      return surface === "admin"
        ? "feature.settings.viewed"
        : "feature.public_page.viewed";
  }
}

function sendToMatomo(
  eventName: AnalyticsEventName,
  payload: AnalyticsEventPayload,
): void {
  const config = getMatomoConfig();
  if (!config || typeof window === "undefined") {
    return;
  }

  const tracker = window._paq ?? [];
  window._paq = tracker;
  const [category, ...rest] = eventName.split(".");
  const action = rest.join(".");
  const name =
    safeName(payload.pageKey) ||
    safeName(payload.route) ||
    safeName(payload.featureArea);
  tracker.push(["trackEvent", category, action, name]);
}

async function sendToBackend(
  eventName: AnalyticsEventName,
  payload: AnalyticsEventPayload,
  schemaVersion = ANALYTICS_BACKEND_EVENT_VERSION,
): Promise<void> {
  if (typeof window === "undefined") {
    return;
  }

  const session = readSessionFromStorage();
  if (!session?.organizationId) {
    return;
  }

  await apiRequest("/analytics/events", {
    method: "POST",
    retry: false,
    json: {
      event_name: eventName,
      schema_version: schemaVersion,
      ...sanitizePayload(payload),
    },
  });
}

export async function trackAnalyticsEvent(
  eventName: AnalyticsEventName,
  payload: AnalyticsEventPayload,
  options?: TrackOptions,
): Promise<void> {
  if (!canTrack()) {
    return;
  }

  const dedupeOnce = options?.dedupeOnce ?? false;
  const dedupeKey =
    payload.dedupeKey ?? (dedupeOnce ? eventKey(eventName, payload) : null);

  if (dedupeKey && !rememberOnce(dedupeKey)) {
    return;
  }

  sendToMatomo(eventName, payload);

  if (options?.sendToBackend === false) {
    return;
  }

  try {
    await sendToBackend(eventName, payload);
  } catch {
    // Fire-and-forget analytics must never block the UI.
  }
}

export async function trackPageView(params: {
  pageKey: string;
  route: string;
  surface: AnalyticsSurface;
  featureArea: AnalyticsFeatureArea;
  locale?: string;
}): Promise<void> {
  await trackAnalyticsEvent(
    resolvePageViewEventName(params.surface, params.featureArea),
    {
      pageKey: params.pageKey,
      route: params.route,
      surface: params.surface,
      featureArea: params.featureArea,
      locale: params.locale,
    },
    { dedupeOnce: true, sendToBackend: params.surface !== "public" },
  );
}

export async function trackActivationEvent(
  eventName:
    | "activation.signup_completed"
    | "activation.organization_created"
    | "activation.first_upload"
    | "activation.first_indexed_document"
    | "activation.first_question"
    | "activation.first_cited_answer",
  payload: Omit<AnalyticsEventPayload, "surface"> & {
    surface: AnalyticsSurface;
  },
): Promise<void> {
  await trackAnalyticsEvent(eventName, payload, { dedupeOnce: true });
}

export async function trackFeatureEvent(
  eventName: AnalyticsEventName,
  payload: Omit<AnalyticsEventPayload, "surface"> & {
    surface: AnalyticsSurface;
  },
): Promise<void> {
  await trackAnalyticsEvent(eventName, payload);
}

export async function trackOnboardingEvent(
  eventName:
    | "onboarding_step_complete"
    | "onboarding_step_skipped"
    | "onboarding_dismissed"
    | "onboarding_tour_started"
    | "onboarding_tour_completed"
    | "onboarding_reset"
    | "onboarding_sample_docs_loaded",
  params?: Record<string, string | number | boolean>,
): Promise<void> {
  if (!canTrack()) {
    return;
  }

  const surface: AnalyticsSurface = "app";
  const payload: AnalyticsEventPayload = {
    surface,
    route: "/organization-onboarding",
    pageKey: "organization-onboarding",
    featureArea: "settings",
    source: eventName,
  };

  if (eventName === "onboarding_sample_docs_loaded") {
    await trackFeatureEvent("feature.settings.viewed", {
      ...payload,
      source: eventName,
    });
    return;
  }

  if (eventName === "onboarding_tour_completed") {
    await trackFeatureEvent("feature.settings.viewed", payload);
    return;
  }

  void params;
  await trackFeatureEvent("feature.settings.viewed", payload);
}

export function hasAnalyticsConsent(): boolean {
  return canTrack();
}

export function readAnalyticsConsentDecisions(): ConsentDecisions | null {
  return readConsentRecord()?.decisions ?? null;
}

File: backend/app/domains/analytics/schemas/analytics.py
from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

AnalyticsEventSurface = Literal["public", "app", "admin"]
AnalyticsFeatureArea = Literal[
    "documents",
    "chat",
    "evaluations",
    "settings",
    "connectors",
    "public",
    "dashboard",
    "admin",
]
AnalyticsSchemaVersion = Literal[1]
AnalyticsEventName = Literal[
    "activation.signup_completed",
    "activation.organization_created",
    "activation.first_upload",
    "activation.first_indexed_document",
    "activation.first_question",
    "activation.first_cited_answer",
    "feature.documents.viewed",
    "feature.documents.uploaded",
    "feature.documents.indexed",
    "feature.dashboard.viewed",
    "feature.chat.viewed",
    "feature.chat.question_submitted",
    "feature.chat.answer_rendered",
    "feature.chat.citation_opened",
    "feature.chat.retrieval_diagnostics_viewed",
    "feature.evaluations.viewed",
    "feature.settings.viewed",
    "feature.connectors.viewed",
    "feature.public_page.viewed",
]

ACTIVATION_EVENT_NAMES = (
    "activation.signup_completed",
    "activation.organization_created",
    "activation.first_upload",
    "activation.first_indexed_document",
    "activation.first_question",
    "activation.first_cited_answer",
)


class AnalyticsDateRange(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    from_date: date = Field(alias="from")
    to_date: date = Field(alias="to")


class AnalyticsEventIngestRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    event_name: AnalyticsEventName
    schema_version: AnalyticsSchemaVersion = 1
    surface: AnalyticsEventSurface
    route: str | None = Field(default=None, max_length=255)
    page_key: str | None = Field(default=None, max_length=128)
    feature_area: AnalyticsFeatureArea | None = None
    entity_id: str | None = Field(default=None, max_length=128)
    entity_type: str | None = Field(default=None, max_length=64)
    status: str | None = Field(default=None, max_length=64)
    method: str | None = Field(default=None, max_length=32)
    count: int | None = Field(default=None, ge=0, le=100_000)
    citation_count: int | None = Field(default=None, ge=0, le=100_000)
    has_citations: bool | None = None
    locale: str | None = Field(default=None, max_length=16)
    source: str | None = Field(default=None, max_length=64)
    dedupe_key: str | None = Field(default=None, max_length=128)

    @field_validator(
        "route", "page_key", "entity_id", "entity_type", "status", "method", "source", "locale"
    )
    @classmethod
    def strip_strings(cls, value: str | None) -> str | None:
        if value is None:
            return None
        trimmed = value.strip()
        return trimmed or None


class AnalyticsEventIngestResponse(BaseModel):
    accepted: bool
    deduped: bool = False
    enabled: bool = True
    event_name: AnalyticsEventName
    schema_version: AnalyticsSchemaVersion = 1


class AnalyticsActivationSummaryResponse(BaseModel):
    signup_completed: int
    organization_created: int
    first_upload: int
    first_indexed_document: int
    first_question: int
    first_cited_answer: int
    funnel_completion_rate: float | None = None


class AnalyticsSummaryResponse(BaseModel):
    organization_id: str
    range: AnalyticsDateRange
    generated_at: datetime
    enabled: bool
    disabled_reason: str | None = None
    total_events: int
    active_users: int
    activation: AnalyticsActivationSummaryResponse
    feature_usage: dict[str, int] = Field(default_factory=dict)
    page_usage: dict[str, int] = Field(default_factory=dict)
    event_counts: dict[str, int] = Field(default_factory=dict)
